### PR TITLE
Fixes #709: Added `FileScan_Search`  Analyzer

### LINKED ISSUE #709: [New analyzer] Integrate new FileScan service
https://filescan.io/api/docs

### PR DESCRIPTION
**Pull request by uzair-ali10**
# Description

Added [FileScan_Search](https://www.filescan.io/api/docs#/reports/search_report_api_reports_search_get) Analyzer

## Related issues
closes #709 

## Type of change

- [x] New feature (non-breaking change which adds functionality).

# Checklist

- [x] The pull request is for the branch develop
- [x] A new analyzer or connector was added, in which case:
    - [x] [Usage](https://github.com/intelowlproject/IntelOwl/blob/master/docs/source/Usage.md) file was updated.
    - [x] If the analyzer/connector requires mocked testing, `_monkeypatch()` was used in it's class to apply the necessary decorators.
- [x] The tests gave 0 errors.
- [x] `Black` gave 0 errors.
- [x] `Flake` gave 0 errors.
# Real World Example

### For IP
Screenshot from 2021-12-07 20-08-59

Screenshot from 2021-12-07 20-09-13
----
### For Domain
Screenshot from 2021-12-07 20-11-35

Screenshot from 2021-12-07 20-11-47

### CODE REVIEW

File: api_app/analyzers_manager/observable_analyzers/filescan_search.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    """FileScan_Search analyzer"""
+
+    base_url: str = "https://www.filescan.io/api"
+
+    def __build_filescan_url(self) -> str:
+        """Builds the URL for the Filescan Search API"""
+        observableName = self.observable_name
+        observableName_bytes = observableName.encode("ascii")
+        base64_bytes = base64.b64encode(observableName_bytes)
+        EncodedObservableName = base64_bytes.decode("ascii")
```
Comment (eshaan7) [inline api_app/analyzers_manager/observable_analyzers/filescan_search.py:26]: do it in one line:

```python3
observable_name_base64 = base64.b64encode(self.observable_name.encode()).decode()
```

File: api_app/analyzers_manager/observable_analyzers/filescan_search.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        endpoint = "reports/search?query={input}"
```
Comment (eshaan7) [inline api_app/analyzers_manager/observable_analyzers/filescan_search.py:27]: the `endpoint` here should be made part of the `base_url`.

File: api_app/analyzers_manager/observable_analyzers/filescan_search.py (hunk 1 above)
Comment (eshaan7) [reply api_app/analyzers_manager/observable_analyzers/filescan_search.py:26]: Also only use snake_case and not PascalCase or camelCase when naming variables.

File: api_app/analyzers_manager/observable_analyzers/filescan_search.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        return f"{self.base_url}/{endpoint.format(input=EncodedObservableName)}"
+
+    def run(self):
+        """Runs the FileScan_Search analyzer"""
+        url = self.__build_filescan_url()
```
Comment (eshaan7) [inline api_app/analyzers_manager/observable_analyzers/filescan_search.py:32]: don't really need the extra `__build_filescan_url` function. Just make it part of `run` only.

File: api_app/analyzers_manager/observable_analyzers/filescan_search.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        response = requests.get(url)
+        if response.status_code != 200:
+            raise AnalyzerRunException(f"FileScan_Search: {response.status_code}")
```
Comment (eshaan7) [inline api_app/analyzers_manager/observable_analyzers/filescan_search.py:35]: you should wrap the HTTP call in a try/except and use `response.raise_for_status()`. See [example](https://github.com/intelowlproject/IntelOwl/blob/3d860e5279e49aee583cd9b4c48e28f4cc894fef/api_app/analyzers_manager/observable_analyzers/shodan.py#L33-L37).

File: api_app/analyzers_manager/observable_analyzers/filescan_search.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        self.result = response.json()
+        return self.result
```
Comment (eshaan7) [inline api_app/analyzers_manager/observable_analyzers/filescan_search.py:37]: single line: `return response.json()`

File: configuration/analyzer_config.json [hunk 6]
```diff
@@ -786,6 +786,26 @@
...
+    "type": "observable",
+    "python_module": "filescan_search.FileScanSearch",
+    "description": "Finds reports and uploaded files by various tokens, like hash, filename, verdict, IOCs etc via [FileScan.io API](https://www.filescan.io/api/docs).",
+    "disabled": false,
+    "external_service": true,
+    "leaks_info": false,
+    "observable_supported": [
+      "url",
+      "domain",
+      "ip",
```
Comment (eshaan7) [inline configuration/analyzer_config.json:799]: also add `hash`.

File: configuration/analyzer_config.json [hunk 7]
```diff
@@ -786,6 +786,26 @@
...
+      "generic"
+    ],
+    "config": {
+      "soft_time_limit": 180,
+      "queue": "long"
+    },
+    "secrets": {},
+    "params": {}
```
Comment (eshaan7) [inline configuration/analyzer_config.json:807]: Add support for these parameters:

- `age`. type: `int`, default: `-1`.
- `method`. type: `str`, default: `and`.
- `page`. type: `int`, default: `1`.
- `page_size`. type: `int`, default: `50`.

File: configuration/analyzer_config.json [hunk 8]
```diff
@@ -786,6 +786,26 @@
...
+    "external_service": true,
+    "leaks_info": false,
+    "observable_supported": [
+      "url",
+      "domain",
+      "ip",
+      "generic"
+    ],
+    "config": {
+      "soft_time_limit": 180,
```
Comment (eshaan7) [inline configuration/analyzer_config.json:803]: change to 60.

File: api_app/analyzers_manager/observable_analyzers/filescan_search.py (hunk 1 above)
Comment (uzair-ali10) [reply api_app/analyzers_manager/observable_analyzers/filescan_search.py:26]: Copy That!!

File: api_app/analyzers_manager/observable_analyzers/filescan_search.py [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+    @classmethod
+    def _monkeypatch(cls):
+        patches = [
+            if_mock_connections(
+                patch(
+                    "requests.get",
+                    return_value=MockResponse({}, 200),
```
Comment (eshaan7) [inline api_app/analyzers_manager/observable_analyzers/filescan_search.py:45]: change to: `MockResponse({'items': [], 'count': 0, 'count_search_params': 1, 'method': 'and'}, 200)

File: api_app/analyzers_manager/observable_analyzers/filescan_search.py [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+class FileScanSearch(ObservableAnalyzer):
+    """FileScan_Search analyzer"""
+
+    base_url: str = "https://www.filescan.io/api"
+
+    def __build_filescan_url(self) -> str:
```
Comment (eshaan7) [inline api_app/analyzers_manager/observable_analyzers/filescan_search.py:21]: move the logic in `__build_filescan_url` to `run` method only.

**Comment by eshaan7** [comment]
Also add entry for this analyzer in the `Usage.md` file.

File: api_app/analyzers_manager/observable_analyzers/filescan_search.py [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+import base64
+import logging
+
+import requests
+
+from api_app.analyzers_manager.classes import ObservableAnalyzer
+from api_app.exceptions import AnalyzerRunException
+from tests.mock_utils import MockResponse, if_mock_connections, patch
+
+logger = logging.getLogger(__name__)
```
Comment (eshaan7) [inline api_app/analyzers_manager/observable_analyzers/filescan_search.py:13]: logger not required here.

File: api_app/analyzers_manager/observable_analyzers/filescan_search.py [hunk 12]
```diff
@@ -0,0 +1,49 @@
+# This file is a part of IntelOwl https://github.com/intelowlproject/IntelOwl
+# See the file 'LICENSE' for copying permission.
+
+import base64
+import logging
+
```
Comment (eshaan7) [inline api_app/analyzers_manager/observable_analyzers/filescan_search.py:6]: redundant empty line.

**Comment by uzair-ali10** [comment]
@eshaan7 
Should I not add extra params in config file?

**Comment by uzair-ali10** [comment]
> > @eshaan7 Should I not add extra params in config file?
> 
> No, because image

Okay, Got it!!

File: docs/source/Usage.md [hunk 13]
```diff
@@ -253,6 +253,7 @@ The following is the list of the available analyzers you can run out-of-the-box.
 * `HashLookupServer_Get_Observable`: check if a md5 or sha1 is available in the database of [known file hosted by CIRCL](https://github.com/adulau/hashlookup-server)
 * `ClamAV`: scan a file via the [ClamAV AntiVirus Engine](https://www.clamav.net/)
 * `Spyse`: Scan domains, IPs, emails and CVEs using Spyse's API. Register [here](https://spyse.com/user/registration).
+* `FileScan_Search`: Finds reports and uploaded files by various tokens, like hash, filename, verdict, IOCs etc via [FileScan.io  API](https://www.filescan.io/api/docs)..
```
Comment (eshaan7) [inline docs/source/Usage.md:256]:
```suggestion
* `FileScan_Search`: Finds reports and uploaded files by various tokens, like hash, filename, verdict, IOCs etc via [FileScan.io  API](https://www.filescan.io/api/docs).
```

File: api_app/analyzers_manager/observable_analyzers/filescan_search.py [hunk 14]
```diff
@@ -0,0 +1,51 @@
+# pylint: disable=raise-missing-from
```
Comment (eshaan7) [inline api_app/analyzers_manager/observable_analyzers/filescan_search.py:1]: why is this line required here ?

**Comment by uzair-ali10** [comment]
Ohh that was added by my Editor, wasn't aware of that

**Comment by eshaan7** [comment]
Thank you, @uzair-ali10 !
